IdentifiabilityODE.mpl

```diff
@@ -557,3 +589,139 @@ CompareDiffVar := proc(dvl, dvr, var_list)
+  # i.e. parameters are states with d/dt = 0
+  visible_states :=  foldl(`union`, op(map(x->indets(rhs(x)) minus {t}, y_eqs))); #select(f -> f in x_zero_vars, ); # map(x->parse(convert(x, string)[..-2]), select(f -> f in x_zero_vars, foldl(`union`, op(map(x->indets(rhs(x)), y_eqs))))):# cat(StringTools[Split](convert(x, string), "_")[1], "_")
+
+  # construct a hash table of "visibility"
+  visibility_table := table([seq(GetStateName(each, x_vars, mu)=current_level, each in visible_states)]):
+  # this is a flag array: if i-th position == 1 then we must differentiat i-th y(t) function 
+  differentiate_ := [seq(1, i=1..nops(y_eqs))]: 
+
+  for j from 1 to s + 1 do
+    # begin differentiation
+    current_level := current_level + 1:
+    continue:=true:
+
+    for i from 1 to m do
+      if differentiate_[i]=1 then 
+        poly_d := numer(lhs(y_eqs[i]) - rhs(y_eqs[i])):
+        poly_d := Differentiate(poly_d, all_vars):
+        leader := MakeDerivative(y_vars[i], j):
+        separant := diff(poly_d, leader):
+        poly_d := simplify(leader - subs(x_eqs, -(poly_d - separant * leader) / separant)):
+        y_eqs[i]:= leader = simplify(poly_d - leader);
+        candidates := select(x-> not (GetOrderVar(x)[1]  in y_vars), indets(y_eqs[i])):
+        if op(map(x->not assigned(visibility_table[GetStateName(x, x_vars, mu)]), candidates)) <> NULL then
+          continue := foldl(`or`, op(map(x->not assigned(visibility_table[GetStateName(x, x_vars, mu)]), candidates))):
+        else
+          continue := false;
+        fi:
+        if continue then
+          differentiate_[i]:=1:
+        else
+          differentiate_[i]:=0:
+        fi:
+        for each in candidates do
+          if not assigned(visibility_table[GetStateName(each, x_vars, mu)]) then 
+              visibility_table[GetStateName(each, x_vars, mu)] := current_level:
+          fi:
+        od;
+      fi:
+    od:
+    if add(k, k in differentiate_)=0 then
+        break:
+    fi:
+  od:
+  return visibility_table;
+end proc:
+
+#===============================================================================
+SubsByDepth := proc(sigma, poly_system, poly_vars, non_id, s, m, x_vars, y_vars, mu, x_eqs, y_eqs, all_vars, {trdegsub:=false})
+#===============================================================================
+  local counting_table_states, min_count, vts, rhs_terms, max_possible,
+        rhs_term, indets_, term, substitutions, each, alg_indep,
+        all_subs, names, selection, other, all_odes, each_ode;
+
+  vts := GetMinLevelBFS(s, m, x_vars, y_vars, mu, x_eqs, y_eqs, all_vars):
+  substitutions := table([]);
+  all_odes := map(x->expand(rhs(x)), select(f->is_diff(lhs(f)), sigma));
+  rhs_terms := []:
+  for each_ode in all_odes do
+    if whattype(each_ode) in [`+`,`*`,`^`] then
+      rhs_terms := [op(rhs_terms), op(each_ode)];
+    end if;
+    if whattype(each_ode) in [function] then
+     rhs_terms := [op(rhs_terms), (each_ode)]; 
+    end if;
+  end do;

```

Comment:
The for-loop seem to extract the functions and parameters appearing in the RHS. Why not to just use `indets` function?

Reply:
I see what you mean, I think I managed to fix, it removed one layer of for looping.